pkg/executor/infoschema_reader.go

```diff
@@ -402,36 +399,38 @@ func (e *memtableRetriever) setDataFromSchemata(ctx sessionctx.Context, schemas
 		rows = append(rows, record)
 	}
 	e.rows = rows
+	return nil
 }
 
-func (e *memtableRetriever) setDataForStatistics(ctx context.Context, sctx sessionctx.Context, schemas []model.CIStr) error {
+func (e *memtableRetriever) setDataForStatistics(ctx context.Context, sctx sessionctx.Context) error {
 	checker := privilege.GetPrivilegeManager(sctx)
-	extractor, ok := e.extractor.(*plannercore.InfoSchemaTablesExtractor)
-	if ok && extractor.SkipRequest {
+	ex, ok := e.extractor.(*plannercore.InfoSchemaStatisticsExtractor)
+	if !ok {
+		return errors.Errorf("wrong extractor type: %T, expected InfoSchemaStatisticsExtractor", e.extractor)
+	}
+	if ex.SkipRequest {
 		return nil
 	}
+	schemas := ex.ListSchemas(e.is)
 	for _, schema := range schemas {
-		if ok && extractor.Filter("table_schema", schema.L) {
-			continue
-		}
-		tables, err := e.is.SchemaTableInfos(ctx, schema)
+		tables, err := ex.ListTables(ctx, e.is, schema)
 		if err != nil {
 			return errors.Trace(err)
 		}
 		for _, table := range tables {
-			if ok && extractor.Filter("table_name", table.Name.L) {
+			if ok && ex.Filter("table_name", table.Name.L) {
```

Comment:
forget to delete this?

Reply:
https://github.com/pingcap/tidb/pull/55144/commits/70405981a1b853e137f12274c94b7e1b3ee35aac

Reply:
for this line, I think we don't need `ex.Filter("table_name", ...)` because line 416 `ex.ListTables` should have done filtering.

Reply:
Removed.